### PR TITLE
{lib}[foss/2021a] AMGX v2.2.0 w/ CUDA 11.3.1

### PR DESCRIPTION
**Pull request by Micket**
(created using `eb --new-pr`)

Disclaimer: Completely untested. Waiting for a GPU node since I suspect the test suite requires it. Not sure how we want to deal with that.

### CODE REVIEW

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ generoso

**Comment by SebastianAchilles** [comment]
@boegelbot please test @ generoso

**Comment by Micket** [comment]
(generoso has no GPUs, so, I suspect it will just fail at the test step)

**Comment by SebastianAchilles** [comment]
> (generoso has no GPUs, so, I suspect it will just fail at the test step)

Yes, that is likely. I will test it on one of the GPU nodes.

**Comment by SebastianAchilles** [comment]
Test report by @sebastianachilles
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
jsfc060 - Linux centos linux 7.9.2009, x86_64, AMD EPYC 7742 64-Core Processor, Python 3.6.8
See https://gist.github.com/a59d747a6013a6415c54094f3c4b59c1 for a full test report.

**Comment by Micket** [comment]
I'm still building magma, test report coming in later

**Comment by Micket** [comment]
I'm still unsure about the patch, and the if-condition (but i really didn't want to make an easyblock)

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0212u17a.bear.cluster - Linux RHEL 8.3, x86_64, Intel(R) Xeon(R) CPU E5-2640 v4 @ 2.40GHz (broadwell), 1 x NVIDIA Tesla P100-PCIE-16GB, 460.32.03, Python 3.6.8
See https://gist.github.com/dc424ec585841a0263a6e9d088addfe8 for a full test report.

File: easybuild/easyconfigs/a/AMGX/AMGX-2.2.0-foss-2021a-CUDA-11.3.1.eb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+dependencies = [
+    ('CUDA', '11.3.1', '', True),
+    ('UCX-CUDA', '1.10.0', versionsuffix),
+    ('magma', '2.6.1', versionsuffix),
+]
+
+if '%(cuda_cc_space_sep)s':
+    configopts = '-DCUDA_ARCH="%(cuda_cc_space_sep)s"'
+
+runtest = "tests -j"
```
Comment (branfosj) [inline easybuild/easyconfigs/a/AMGX/AMGX-2.2.0-foss-2021a-CUDA-11.3.1.eb:35]:
```
== 2021-10-04 12:55:00,867 run.py:582 INFO cmd " make tests -j " exited with exit code 0 and output:
make: Nothing to be done for 'tests'.
```

**Comment by akesandgren** [comment]
Test report by @akesandgren
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
b-cn1113.hpc2n.umu.se - Linux Ubuntu 20.04, x86_64, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 3.8.10
See https://gist.github.com/2837869975bd31c449802a96aa15517b for a full test report.

**Comment by akesandgren** [comment]
How did you get those builds to not fail with ICE's???
I get:
```
/hpc2n/eb/software/CUDA/11.3.1/bin/nvcc /scratch/eb-ake/AMGX/2.2.0/foss-2021a-CUDA-11.3.1/AMGX-2.2.0/core/src/solvers/user_solver.cu -c -o /scratch/eb-ake/AMGX/2.2.0/foss-2021a-CUDA-11.3.1/easybuild_obj/core/CMakeFiles/amgx_core.dir/src/solvers/./amgx_core_generated_user_solver.cu.o -ccbin /hpc2n/eb/software/GCCcore/10.3.0/bin/gcc -m64 -Xcompiler ,\"-O2\",\"-ftree-vectorize\",\"-march=native\",\"-fno-math-errno\",\"-fopenmp\",\"-Wno-terminate\",\"-static-libgcc\",\"-fopenmp\",\"-DRAPIDJSON_DEFINED\",\"-DAMGX_WITH_MPI\",\"-O3\",\"-DNDEBUG\" -Xcompiler=-rdynamic -Xcompiler=-fPIC -Xcompiler=-fvisibility=default -DDISABLE_MIXED_PRECISION -DCUSPARSE_GENERIC_INTERFACES -DCUSPARSE_USE_GENERIC_SPGEMM -O3 -DNDEBUG -std=c++14 --Werror cross-execution-space-call -DNVTX_RANGES -DNVCC -I/scratch/eb-ake/AMGX/2.2.0/foss-2021a-CUDA-11.3.1/AMGX-2.2.0/../../thrust -I/scratch/eb-ake/AMGX/2.2.0/foss-2021a-CUDA-11.3.1/AMGX-2.2.0/core/include -I/scratch/eb-ake/AMGX/2.2.0/foss-2021a-CUDA-11.3.1/AMGX-2.2.0/core/../base/include -I/hpc2n/eb/software/CUDA/11.3.1/include -I/scratch/eb-ake/AMGX/2.2.0/foss-2021a-CUDA-11.3.1/AMGX-2.2.0/external/rapidjson/include -I/cvmfs/ebsw.hpc2n.umu.se/amd64_ubuntu2004_bdw/software/OpenMPI/4.1.1-GCC-10.3.0/include -I/hpc2n/eb/software/OpenMPI/4.1.1-GCC-10.3.0/include
/cvmfs/ebsw.hpc2n.umu.se/amd64_ubuntu2004_bdw/software/GCCcore/10.3.0/include/c++/10.3.0/chrono: In substitution of template<class _Rep, class _Period> template<class _Period2> using __is_harmonic = std::__bool_constant<(std::ratio<((_Period2::num / std::chrono::duration<_Rep, _Period>::_S_gcd(_Period2::num, _Period::num)) * (_Period::den / std::chrono::duration<_Rep, _Period>::_S_gcd(_Period2::den, _Period::den))), ((_Period2::den / std::chrono::duration<_Rep, _Period>::_S_gcd(_Period2::den, _Period::den)) * (_Period::num / std::chrono::duration<_Rep, _Period>::_S_gcd(_Period2::num, _Period::num)))>::den == 1)> [with _Period2 = _Period2; _Rep = _Rep; _Period = _Period]:
/cvmfs/ebsw.hpc2n.umu.se/amd64_ubuntu2004_bdw/software/GCCcore/10.3.0/include/c++/10.3.0/chrono:473:154:   required from here
/cvmfs/ebsw.hpc2n.umu.se/amd64_ubuntu2004_bdw/software/GCCcore/10.3.0/include/c++/10.3.0/chrono:428:27: internal compiler error: Segmentation fault
  428 |  _S_gcd(intmax_t __m, intmax_t __n) noexcept
      |                           ^~~~~~
0xcbbdaf crash_signal
        ../../gcc/toplev.c:328
0x7b1e7d tsubst(tree_node*, tree_node*, int, tree_node*)
        ../../gcc/cp/pt.c:15310
...
```

**Comment by branfosj** [comment]
@akesandgren Have you built GCCcore since https://github.com/easybuilders/easybuild-easyconfigs/pull/13310 was merged?

**Comment by akesandgren** [comment]
Perhaps not.... rebuilding it...

**Comment by akesandgren** [comment]
Test report by @akesandgren
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
b-cn1113.hpc2n.umu.se - Linux Ubuntu 20.04, x86_64, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 3.8.10
See https://gist.github.com/b93f8c1d2fcd3602b5a51d7db5fafe73 for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
vera54-1 - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz, Python 3.6.8
See https://gist.github.com/bae8987fed8ad81d04c2cc180adca76e for a full test report.

**Comment by Micket** [comment]
So, one can build
```
make amgx_tests_launcher
```
and then run that. It seems to be running things in the GPU, but, it all fails for me. 

```
Spawning new worker
[   1/ 183] AggregatesCoarseGeneratorTest        dDDI : [PASSED]
[   2/ 183] AggregatesCoarseGeneratorTest        dDFI : [PASSED]
[   3/ 183] AggregatesCoarseGeneratorTest        dFFI : [PASSED]
[   4/ 183] AggregatesCoarseningFactor   dDDI : [FAILED]
Spawning new worker
Launching test AggregatesCoarseningFactor dDDI in the new process for the second time
[   1/ 180] AggregatesCoarseningFactor   dDDI : [FAILED]
[   2/ 180] AggregatesCoarseningFactor   dDFI : [FAILED]
Spawning new worker
Launching test AggregatesCoarseningFactor dDFI in the new process for the second time
[   1/ 179] AggregatesCoarseningFactor   dDFI : [FAILED]
[   2/ 179] AggregatesCoarseningFactor   dFFI : ^CCaught signal 2 - SIGINT (interrupt)
```
I stopped it after 15 hours. What do you guys think?

**Comment by akesandgren** [comment]
I've installed it and thrown it at the user who requested it.

**Comment by akesandgren** [comment]
their ci/test.sh says
```
# WIP: test_launcher is allowed to fail; not all tests pass
./tests/amgx_tests_launcher
```
so perhaps all is fine?

File: easybuild/easyconfigs/a/AMGX/AMGX-2.2.0-foss-2021a-CUDA-11.3.1.eb [hunk 2]
```diff
@@ -32,10 +32,8 @@ dependencies = [
 if '%(cuda_cc_space_sep)s':
     configopts = '-DCUDA_ARCH="%(cuda_cc_space_sep)s"'
 
-runtest = "tests -j"
-
 sanity_check_paths = {
-    'files': ['include/amgx_c.h', 'lib/libamgx.a', 'lib/libamgxsh.%s' % SHLIB_EXT],
+    'files': ['include/amgx_c.hx', 'lib/libamgx.a', 'lib/libamgxsh.%s' % SHLIB_EXT],
```
Comment (akesandgren) [inline easybuild/easyconfigs/a/AMGX/AMGX-2.2.0-foss-2021a-CUDA-11.3.1.eb:36]: I can't find any include/amgx_c.hx file in our install

**Comment by akesandgren** [comment]
Test report by @akesandgren
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
b-cn1312.hpc2n.umu.se - Linux Ubuntu 20.04, x86_64, Intel(R) Xeon(R) CPU E5-2690 v4 @ 2.60GHz, Python 3.8.10
See https://gist.github.com/21292553fdf49f0f9ddf6df6bd1ed748 for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
vera54-1 - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6130 CPU @ 2.10GHz, Python 3.6.8
See https://gist.github.com/cbf86d23561048de81f99fc381cf41a8 for a full test report.

**Review by jfgrimm — COMMENTED** [review]
As decided in issue #16330, we have deprecated the use of `True` to signify a system-toolchain dependency (#16384), in favour of the more intuitive `SYSTEM` template constant. Due to the change in the test suite, please run `eb --sync-pr-with-develop 14094` and update the PR to use `SYSTEM` instead.

**Comment by Micket** [comment]
AMGX 2.3.0 is already out, and revamped their CMakeLists properly, so this is outdated. This is replaced by #16560

**Pull request closed without merging**